static/js/choropleth/choropleth.ts

```diff
@@ -0,0 +1,178 @@
+
+/**
+ * Handles the creation of d3 map from geojson file.
+ * @param {json} geojson entire geoJson map.
+ */
+function generateGeoMap(geojson, plt_info) {
+    // Combine path elements from D3 content.
+    var mapContent = d3.select("#main-pane g.map")
+        .selectAll("path")
+        .data(geojson.features);
+
+    // Build chart display options.
+    var colorScale = d3.scaleLinear()
+      .domain(plt_info['domain'])
+      .range(plt_info['palette']);    
+
+    // Scale and center the map.
+    var svg_container = document.getElementById("map_container")
+    var projection = d3.geoAlbers().fitSize([
+        svg_container.clientWidth,
+        svg_container.clientHeight],
+        geojson);
+    var geomap = d3.geoPath().projection(projection);
+
+    // Build map objects.
+    mapContent.enter()
+        .append("path")
+        .attr("d", geomap)
+        // Add CSS class to each path for border outlining.
+        .attr("class", "border")
+        .attr("fill", function (d) {
+            if (d.properties.hasOwnProperty('value')) {
+                return colorScale(d.properties.value)
+            } else {
+                return "gray";
+            }
+          })
+        // Add various country level event handlers.
+        .on("mouseover", handleMapHover)
+        .on("mouseleave", mouseLeave)
+        .on("click", handleMapClick);
+}
+
+/**
+ * Capture hover event on country and displays relevant information.
+ * @param {json} country is the geoJson content for the hovered country.
+ */
+function handleMapHover(country) {
+    // Display statistical variable information on hover.
+    let name = country.properties.name;
+    if (country.properties.hasOwnProperty('value')) {
+        document.getElementById("display").innerHTML = name + " - " +
+            country.properties.value;
+    } else {
+        document.getElementById("display").innerHTML = name + " - No Value";
+    }
+    
+    // Highlight selected country in black on hover.
+    d3.select(this).style("stroke", "black");
+  }
+  
+  /**
+   * Clears output after leaving a country.
+   */
+  function mouseLeave() {
+    // Remove hover text.
+    document.getElementById("display").innerHTML = "";
+  
+    // Remove country display effect.
+    d3.select(this).style("stroke", "white");
+  }
+  
+  /**
+   * Capture click event on country and redirects
+   * user to that page in the browser.
+   * @param {json} country is the geoJson content for the clicked country.
+   */
+  function handleMapClick(geo) {
+    if (geo.properties.hasSublevel) {
+      redirectToGeo(geo.properties.geoId)
+    } else {
+      alert("This geo has no further sublevels!")
+    }
+  }
+
+  function redirectToGeo(geoId) {
+    var base_url = document.location.origin + document.location.pathname + '?'
+    base_url += "statVar=" + url.searchParams.get("statVar");
+    base_url += "&perCapita=" + url.searchParams.get("perCapita")
+    base_url += "&place=" + geoId
+    base_url += "&bc=" 
```

Comment:
Is this handled by the server?

Reply:
The breadcrumbs are ignored.

Reply:
The breadcrumbs are not sent to the server*